portalnet/src/overlay_service.rs

```diff
@@ -2402,7 +2402,34 @@ where
             peer_timeout: self.query_peer_timeout,
         };
 
-        let closest_enrs = self.closest_connected_nodes(&target_key, query_config.num_results);
+        let mut closest_enrs = self.closest_connected_nodes(&target_key, query_config.num_results);
+        if closest_enrs.is_empty() {
+            // If there are no connected nodes, use the closest nodes regardless of connectivity.
+            closest_enrs = self
+                .kbuckets
+                .write()
```

Comment:
Hm, I was wondering why this needs to be a write lock when it seems to be a read-only query. I'm assuming you tried a read lock first. Assuming you did, I think the `&mut self` on `closest_values` is why it's required. Though I can't tell from looking at the source why it couldn't be a `&self` instead:
https://github.com/sigp/discv5/blob/1439decd4e7d7c9de78ef61b5d67be3fee688510/src/kbucket.rs#L650-L670

Might be worth seeing if it's as easy as removing the `mut` in a draft PR against discv5. (To be clear, I *don't* mean that getting the discv5 change should block this PR).

Reply:
I wondered the same. 

I'll see if that discv5 change breaks anything after merging.